### PR TITLE
lib/gis: allow setting vector/raster processing custom tmp directory via shell env variable

### PR DESCRIPTION
**Pull request by tmszi**
With this small patch it is possible set 'GRASS_RASTER_TMPDIR_MAPSET' shell env variable (zero value or any directory) issue #893, and 'GRASS_VECTOR_TMPDIR_MAPSET' (originally only [zero value ](https://grass.osgeo.org/grass79/manuals/variables.html) could be set) to any other directory. A subdirectory (`.tmp/hostname/`) of files is always created in the setted directory. Could @neteler test it, please?

**Simple test** (nc_basic_spm_grass7 location):

Raster process:
```
g.region rast=elevation

v.random output=elev_pts npoints=10000 column=z
v.what.rast map=elev_pts raster=elevation column=z

# make custom raster tmp dir
mkdir /tmp/gg_raster_tmp

export GRASS_RASTER_TMPDIR_MAPSET=/tmp/gg_raster_tmp

# start another virtual terminal tab
watch ls -la "/tmp/gg_raster_tmp/.tmp/`hostname`/"

# check directory content, previous step
v.surf.rst input=elev_pts zcolumn=z elevation=elev_test slope=slope_test aspect=aspect_test
```

Vector process:

```
# make custom raster tmp dir
mkdir /tmp/gg_vector_tmp

export GRASS_VECTOR_TMPDIR_MAPSET=/tmp/gg_vector_tmp

# start another virtual terminal tab
watch ls -la "/tmp/gg_vector_tmp/.tmp/`hostname`/vector"

# elev_pts vector map is from previous raster process example
# check directory content, previous step
v.buffer input=elev_pts output=elev_pts_buffer distance=5
```

### CODE REVIEW

**Comment by tmszi** [comment]
In the [929ca2f](https://github.com/OSGeo/grass/commit/929ca2f28abd7695c54a9c326d69d065c70b8441)  replace [strsep()](https://www.man7.org/linux/man-pages/man3/strsep.3.html) function (which is not a POSIX or GNU/Linux standard function, and on MS Windows OS throw this warning message `implicit declaration of function ‘strsep’`) with custom [strsep_()](https://github.com/OSGeo/grass/commit/929ca2f28abd7695c54a9c326d69d065c70b8441#diff-6585db6536188f8bbf1236078559636eR201) function. 

Rebase in the [d20ec70](https://github.com/OSGeo/grass/commit/d20ec709e6b3d5d97071f47b5b9ba99535694a27).

**Comment by neteler** [comment]
Amazing, @tmszi !

I tested both, and for fun also this computation:

```
g.region res=1 -p
r.mapcalc "diff = elev_test - elevation"
```
I wonder if any other tests are needed, maybe @metzm has an idea?

## Special cases

Some potential corner cases come to mind (I didn't check the code yet):
- r.watershed: in "disk swap memory" mode, it might also use this new variable
- r.terraflow: directory can be specified but what is the default value?
- r.viewshed: directory can be specified but what is the default value?
- t.rast.export: path to the work directory, default is /tmp
- t.rast.import: path to the work directory, default is /tmp
- t.vect.export: path to the work directory, default is /tmp
- t.vect.import: path to the work directory, default is /tmp
- more?

Thanks for your hard work!

**Comment by tmszi** [comment]
**Special cases**

You are welcome. I tested special cases.

```
# respect GRASS_RASTER_TMPDIR_MAPSET shell env var
r.watershed -m elevation=elevation accumulation=accumulation_test tci=tci_test spi=spi_test drainage=drainage_test basin=basin_test length_slope=length_slope_test threshold=10000 --o

# respect GRASS_RASTER_TMPDIR_MAPSET shell env var
r.terraflow elevation=elevation filled=elev_filled_test direction=dir_test swatershed=swatershed_test --o

# respect GRASS_RASTER_TMPDIR_MAPSET shell env var
r.viewshed input=elevation output=viewshed_test coordinates=631969.1629955947,226388.76651982378 --o

# respect GRASS_RASTER_TMPDIR_MAPSET shell env var
# 3D raster interpolation
# elev_pts vector points map from previous first simple raster example
v.to.3d input=elev_pts output=elev_pts_3d column=z
v.db.addcolumn map=elev_pts_3d columns="w double precision"
v.db.update map=elev_pts_3d column=w value=50
g.region res3=100 t=1000 b=0
v.vol.rst input=elev_pts_3d elevation=elev_test gradient=gradient_test aspect_horizontal=aspect_horizontal_test aspect_vertical=aspect_vertical_test ncurvature=ncurvature_test gcurvature=gcurvature_test mcurvature=mcurvature_test wcolumn=w
```

I tested add web services layer WMS (wxGUI Add web service layer dialog), also respect `GRASS_RASTER_TMPDIR_MAPSET` shell env var.

**Comment by neteler** [comment]
I went on with testing, a big job with Sentinel-2 time series is currently running on a new machine.

Doing so I saw that when the `GRASS_RASTER_TMPDIR_MAPSET` / `GRASS_VECTOR_TMPDIR_MAPSET` env variables aren't set, this error occurs:

```
g.region vector=south_bbox -pa res=10
ERROR: Unable to make mapset element .tmp/calc
       (/scratch_ceph/s2_cloudfree/grassdata_tmp_raster/.tmp): No such
       file or directory
```

I wonder if this error could be caught early enough, maybe in libgis?

**Comment by tmszi** [comment]
> I went on with testing, a big job with Sentinel-2 time series is currently running on a new machine.
> 
> Doing so I saw that when the `GRASS_RASTER_TMPDIR_MAPSET` / `GRASS_VECTOR_TMPDIR_MAPSET` env variables aren't set, this error occurs:
> 
> ```
> g.region vector=south_bbox -pa res=10
> ERROR: Unable to make mapset element .tmp/calc
>        (/scratch_ceph/s2_cloudfree/grassdata_tmp_raster/.tmp): No such
>        file or directory
> ```
> 
> I wonder if this error could be caught early enough, maybe in libgis?

I can't reproduce this error (nc_basic_spm_grass7 location). My simple test:

```
grass79

echo $LOCATION
/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT

# GRASS_VECTOR_TMPDIR_MAPSET / GRASS_RASTER_TMPDIR_MAPSET shell env vars aren't setted
# Set region according vector map
g.region -pa vect=geology res=10
projection: 99 (Lambert Conformal Conic)
zone:       0
datum:      nad83
ellipsoid:  a=6378137 es=0.006694380022900787
north:      318150
south:      10850
west:       123950
east:       930200
nsres:      10
ewres:      10
rows:       30730
cols:       80625
cells:      2477606250

# Set GRASS_VECTOR_TMPDIR_MAPSET shell env var to custom tmp dir
export GRASS_VECTOR_TMPDIR_MAPSET=/tmp/gg_tmp_vector/

# Set GRASS_RASTER_TMPDIR_MAPSET shell env var to custom tmp dir
export GRASS_RASTER_TMPDIR_MAPSET=/tmp/gg_tmp_raster/

# Set region according vector map again
g.region -pa vect=geology res=10
projection: 99 (Lambert Conformal Conic)
zone:       0
datum:      nad83
ellipsoid:  a=6378137 es=0.006694380022900787
north:      318120
south:      10840
west:       123960
east:       930200
nsres:      10
ewres:      10
rows:       7682
cols:       20156
cells:      154838392

# Unset  GRASS_VECTOR_TMPDIR_MAPSET shell env var
unset GRASS_VECTOR_TMPDIR_MAPSET

# Unset  GRASS_RASTER_TMPDIR_MAPSET shell env var
unset GRASS_RASTER_TMPDIR_MAPSET

# Set region according vector map again
g.region -pa vect=geology res=10
projection: 99 (Lambert Conformal Conic)
zone:       0
datum:      nad83
ellipsoid:  a=6378137 es=0.006694380022900787
north:      318150
south:      10800
west:       123950
east:       930200
nsres:      10
ewres:      10
rows:       6147
cols:       16125
cells:      99120375
```

Could you provide more information to reproduce this error, please?

**Comment by tmszi** [comment]
**This patch doesn't appear to work as expected**. I found another error during my testing if `GRASS_RASTER_TMPDIR_MAPSET=/tmp/gg_tmp_raster/` and /tmp dir is located on another device partition than LOCATION dir (every raster operation fail, previous tests).

Raster process:

```
# /tmp dir is located on another device partition
export GRASS_RASTER_TMPDIR_MAPSET=/tmp/gg_raster_tmp

v.surf.rst input=elev_pts zcolumn=z elevation=elev_test slope=slope_test aspect=aspect_test --o
Reading features from vector map ...
Ignoring 465 points (too dense)
Processing all selected output files will require
654.66 KB of disk space for temp files.
Processing segments in parallel...
Starting parallel work
 100%
WARNING: Unable to rename null file
         '/tmp/gg_raster_tmp//.tmp/gentoo-gnu-linux/29105.8' to
         '/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT/cell_misc/elev_test/nullcmpr':
         Invalid cross-device link
WARNING: Unable to rename cell file
         '/tmp/gg_raster_tmp//.tmp/gentoo-gnu-linux/29105.7' to
         '/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT/fcell/elev_test':
         Invalid cross-device link
WARNING: Unable to write quant rules: raster map <elev_test> is integer
WARNING: Unable to rename null file
         '/tmp/gg_raster_tmp//.tmp/gentoo-gnu-linux/29105.10' to
         '/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT/cell_misc/slope_test/nullcmpr':
         Invalid cross-device link
WARNING: Unable to rename cell file
         '/tmp/gg_raster_tmp//.tmp/gentoo-gnu-linux/29105.9' to
         '/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT/fcell/slope_test':
         Invalid cross-device link
WARNING: Unable to write quant rules: raster map <slope_test> is integer
WARNING: Unable to rename null file
         '/tmp/gg_raster_tmp//.tmp/gentoo-gnu-linux/29105.12' to
         '/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT/cell_misc/aspect_test/nullcmpr':
         Invalid cross-device link
WARNING: Unable to rename cell file
         '/tmp/gg_raster_tmp//.tmp/gentoo-gnu-linux/29105.11' to
         '/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT/fcell/aspect_test':
         Invalid cross-device link
WARNING: Unable to write quant rules: raster map <aspect_test> is integer
WARNING: Unable to write quant rules: raster map <elev_test> is integer
WARNING: Unable to write quant rules: raster map <slope_test> is integer
WARNING: Unable to write quant rules: raster map <aspect_test> is integer
v.surf.rst complete.
```

Vector process:

```
# /tmp dir is located on nother device partition
# set  GRASS_RASTER_TMPDIR_MAPSET shell env var too before vector operation too
export GRASS_RASTER_TMPDIR_MAPSET=/tmp/gg_raster_tmp
export GRASS_VECTOR_TMPDIR_MAPSET=/tmp/gg_vector_tmp/

v.buffer input=elev_pts output=elev_pts_buffer distance=5 --o
ERROR: Unable to make mapset element .tmp/gentoo-gnu-linux/vector/tmp_29617
       (/tmp/gg_vector_tmp/.tmp): No such file or directory
WARNING: Unable to rename directory
         '/home/tomas/grassdata/nc_basic_spm_grass7/PERMANENT/vector/elev_pts_buffer0'
         to '/tmp/gg_raster_tmp//.tmp/gentoo-gnu-linux/29617.4'
```

**Comment by neteler** [comment]
> **This patch doesn't appear to work as expected**. I found another error during my testing if `GRASS_RASTER_TMPDIR_MAPSET=/tmp/gg_tmp_raster/` and /tmp dir is located on another device partition than LOCATION dir (every raster operation fail, previous tests).

This seems to be exactly my problem: I am using different network (NFS) drives + "local" (CEPH) drive.
I guess it is this function causing troubles in this context:
https://github.com/OSGeo/grass/blob/2d24a42cd4b6c66ab81348aa61240e784df39032/lib/gis/rename.c#L40

**Comment by metzm** [comment]
The function `G_rename_file()`
> 
> https://github.com/OSGeo/grass/blob/2d24a42cd4b6c66ab81348aa61240e784df39032/lib/gis/rename.c#L40

has a fallback in case `rename()` is not possible. The problem must be somewhere else.

The above mentioned error
```
ERROR: Unable to make mapset element .tmp/gentoo-gnu-linux/vector/tmp_29617
       (/tmp/gg_vector_tmp/.tmp): No such file or directory
```
needs to be fixed first. Maybe the path given with GRASS_VECTOR_TMPDIR_MAPSET must exist before the env var is set?

File: lib/gis/file_name.c [hunk 1]
```diff
@@ -79,16 +80,53 @@ char *G_file_name_misc(char *path,
 */
 char *G_file_name_tmp(char *path,
                       const char *element,
-                      const char *name, const char *mapset)
+                      const char *name, const char *mapset,
+                      const char *type)
 {
     const char *env, *tmp_path;
+    const char path_sep =
```
Comment (metzm) [inline lib/gis/file_name.c:87]: Use `G_convert_dirseps_*()` functions.

File: lib/gis/file_name.c [hunk 2]
```diff
@@ -79,16 +80,53 @@ char *G_file_name_misc(char *path,
...
+    #ifdef _WIN32
+        '\\';
+    #else
+        '/';
+    #endif
 
     tmp_path = NULL;
-    env = getenv("GRASS_VECTOR_TMPDIR_MAPSET");
-    if (env && strcmp(env, "0") == 0) {
```
Comment (metzm) [inline lib/gis/file_name.c:88]: You probably mean `strcmp(env, "\0") != 0)

File: lib/gis/file_name.c [hunk 3]
```diff
@@ -79,16 +80,53 @@ char *G_file_name_misc(char *path,
...
-    
+    else if (element) {
+        int c = 1;
+        char *found;
+        char *e;
+        e = strdup(element);
+        while ((found = strsep_(&e, &path_sep))) {
+            if (c == 3 && strcmp(found, "vector") == 0) {
+                env = getenv("GRASS_VECTOR_TMPDIR_MAPSET");
+                if (env && strcmp(env, "0") == 0)
```
Comment (metzm) [inline lib/gis/file_name.c:110]: as above

File: lib/gis/file_name.c [hunk 4]
```diff
@@ -79,16 +80,53 @@ char *G_file_name_misc(char *path,
...
+        '/';
+    #endif
 
     tmp_path = NULL;
-    env = getenv("GRASS_VECTOR_TMPDIR_MAPSET");
-    if (env && strcmp(env, "0") == 0) {
+    if (type && strcmp(type, "vector") == 0) {
+      env = getenv("GRASS_VECTOR_TMPDIR_MAPSET");
+      if (env && strcmp(env, "0") == 0)
         tmp_path = getenv("TMPDIR");
```
Comment (metzm) [inline lib/gis/file_name.c:98]: Why use TMPDIR?

File: lib/gis/open.c [hunk 5]
```diff
@@ -21,6 +21,7 @@
 
 #include <grass/gis.h>
 #include <grass/glocale.h>
+#include <grass/vect/dig_defines.h>
```
Comment (metzm) [inline lib/gis/open.c:24]: This is not allowed. Vector headers can only be included by vector libs or libs depending on vector libs or modules.

File: include/defs/gis.h [hunk 6]
```diff
@@ -352,7 +352,8 @@ void G_init_logging(void);
 char *G_file_name(char *, const char *, const char *, const char *);
 char *G_file_name_misc(char *, const char *, const char *, const char *,
 		       const char *);
-char *G_file_name_tmp(char *, const char *, const char *, const char *);
+char *G_file_name_tmp(char *, const char *, const char *, const char *,
```
Comment (metzm) [inline include/defs/gis.h:355]: This change requires an increase of the GRASS major version.
Why not add GRASS_VECTOR_TMPDIR_MAPSET to `G_file_name_tmp()`, update this fn and update relevant code in lib/raster? Most of the other changes in lib/gis could then be removed.

**Comment by neteler** [comment]
I am really interested in getting this working. Perhaps we can resume updating and testing? Happy to support here...

**Comment by tmszi** [comment]
Implemented by PR #1786.
